### PR TITLE
Restructure Security chapter

### PR DESCRIPTION
**Pull request by tarekoraby**
This PR:

1. Prioritizes the description of Spring + Flow's built-in view-based access control mechanism to set up security.
2. Move all other security discussions under an "Advanced Security Topics" sub-menu (so as not to distract the typical user whose need is to just enable security).

This is related to https://github.com/vaadin/docs/issues/967.

### CODE REVIEW

File: articles/flow/security/advanced-topics/security-practices.adoc [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+= Security Practices at Vaadin
+
+== Releasing Security Patches
+
+Security fixes are implemented as fast as possible and released for all currently supported versions.
+The fix is mentioned in the release notes, and we also send a separate security notification email to all our registered users, explaining the issue and how to fix it (typically by updating to a new maintenance version).
+
+== How Users Can Report Security Issues
+
+If a developer or user finds a potential security issue, they can report it directly to `security@vaadin.com`.
```
Comment (knoobie) [inline articles/flow/security/advanced-topics/security-practices.adoc:17]: Should the mail address be replaced with a link to the dedicated security page? https://vaadin.com/security

Comment (tarekoraby) [reply articles/flow/security/advanced-topics/security-practices.adoc:17]: Well, the security page is mentioning the email. But perhaps it's worth mentioning both the email, and the security page for more details.

File: articles/flow/security/advanced-topics/architecture.adoc [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+On the server, the developer can load the user object again, set the new name, and store the object.
+
+As the example shows, the developer can safely handle confidential data in their code, even when dealing with UI code.
+The framework only stores data that the developer gives it explicitly.
+Only data that should be visible on the client is actually sent there.
+
+== Application State
+
+In a Vaadin Flow application, the server is always aware of your application's state.
+This means that, like client-side applications, the server is aware of what is currently visible on the end user's screen.
```
Comment (dougjewsbury) [inline articles/flow/security/advanced-topics/architecture.adoc:46]: Should that be "like" or "unlike" client-side applications?

File: articles/flow/security/advanced-topics/architecture.adoc [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+// tag::data-validation[]
+== Data Validation
+
+In a Vaadin application, the data binding API supports data validation on the server, which cannot be bypassed by client-side attacks.
+Vaadin components do support client-side validation to increase the responsiveness of the application, but the developer should be aware that these should be used purely for convenience, since they are easily circumvented in the browser.
+
+As with other web applications, all data coming from the client should always be validated once it reaches the server.
+It is not safe to rely only on client-side validation.
+Vaadin provides a set of pre-created server side validators for this purpose.
```
Comment (dougjewsbury) [inline articles/flow/security/advanced-topics/architecture.adoc:71]:
```suggestion
Vaadin provides a set of pre-created server-side validators for this purpose.
```

File: articles/flow/security/advanced-topics/securing-plain-java-app.adoc [hunk 4]
```diff
@@ -1,66 +1,43 @@
...
 layout: page
 ---
 
-= Security Best Practices
+= Securing Plain Java Applications
 
-// tag::authentication-authorization[]
-== Authentication and Authorization
+As described in the <<{articles}/flow/security/enabling-security#, enabling security>> chapter, Vaadin Flow comes with built-in security helpers that are most convenient to implement with Spring Boot and Spring Security.
+While recommended, using Spring Security is not mandatory, however.
```
Comment (dougjewsbury) [inline articles/flow/security/advanced-topics/securing-plain-java-app.adoc:11]:
```suggestion
Although it is recommended to use Spring Security, it is not mandatory.
```

File: articles/flow/security/advanced-topics/securing-plain-java-app.adoc [hunk 5]
```diff
@@ -1,66 +1,43 @@
...
 - A security realm.
 - A log-in view.
 - Log-out capability.
 - A [interfacename]`VaadinServiceInitListener` that adds [classname]`ViewAccessChecker` as the [classname]`BeforeEnterListener` of all UIs.
 - A Service Provider under `META-INF/services` to load above the custom [classname]`ServiceInitListener` via the Java SPI loading mechanism.
 
-=== Configuring a Security Realm
+== Configuring a Security Realm
 
 A _security realm_ is a mechanism for storing and mapping the users with their passwords and roles.
```
Comment (dougjewsbury) [inline articles/flow/security/advanced-topics/securing-plain-java-app.adoc:34]:
```suggestion
A _security realm_ is a mechanism for storing and mapping users to their passwords and roles.
```

File: articles/flow/security/advanced-topics/securing-plain-java-app.adoc [hunk 6]
```diff
@@ -236,7 +213,7 @@ Now it is ready to intercept attempts to enter all views.
 
 However, the above class still needs to be loaded, so we recommend following the instructions in the next step.
```
Comment (dougjewsbury) [inline articles/flow/security/advanced-topics/securing-plain-java-app.adoc:214]:
```suggestion
However, this class still needs to be loaded, so we recommend following the instructions in the next step.
```

File: articles/flow/security/advanced-topics/security-practices.adoc [hunk 7]
```diff
@@ -0,0 +1,38 @@
+---
+title: Security Practices at Vaadin
+description: Provides an overview to the steps Vaadin takes to ensure your application remains secure.
```
Comment (dougjewsbury) [inline articles/flow/security/advanced-topics/security-practices.adoc:3]:
```suggestion
description: Provides an overview of the steps Vaadin takes to ensure your application remains secure.
```

File: articles/flow/security/enabling-security.adoc [hunk 8]
```diff
@@ -1,358 +1,338 @@
...
+order: 10
+layout: page
+---
+
+
+= Securing Spring Boot Applications
+
+Vaadin is fully compatible with the most-popular security solutions in the Java ecosystem, including but not limited to Spring Security, JAAS, and Apache Shiro.
+And while you can choose the security framework you want to use, Vaadin Flow comes with built-in security helpers that are most convenient to implement with Spring Security.
+Using those built-in helpers with Spring Security makes it easier, faster, and safer to secure your Vaadin applications than if you were to take use of the Spring Security framework directly, or if you were to use another security framework.
```
Comment (dougjewsbury) [inline articles/flow/security/enabling-security.adoc:13]:
```suggestion
And although you can choose the security framework you want to use, Vaadin Flow comes with built-in security helpers that are most convenient to implement with Spring Security.
Using these built-in helpers with Spring Security makes it easier, faster, and safer to secure your Vaadin applications than if you were use the Spring Security framework directly, or if you were to use another security framework.

```

File: articles/flow/security/enabling-security.adoc [hunk 9]
```diff
@@ -1,358 +1,338 @@
...
+This view-based access control mechanism enables you to fully secure views in Vaadin Flow applications in a flexible way, based on different access level annotations.
+Specifically, the view-based access control mechanism uses the [annotationname]`@AnonymousAllowed`, [annotationname]`@PermitAll`, [annotationname]`@RolesAllowed`, and [annotationname]`@DenyAll` annotations on view classes to define the access control rules.
+
+To enable the mechanism in a Vaadin Flow Spring Boot application without any security, add the following to the project (if they do not exist already):
+
+- A Login view.
+- Spring Security dependencies.
+- Log-out capability.
+- A security configuration class that extends [classname]`VaadinWebSecurityConfigurerAdapter`.
+- Annotate the view classes with the [annotationname]`@AnonymousAllowed`, [annotationname]`@PermitAll`, or [annotationname]`@RolesAllowed` annotations.
```
Comment (dougjewsbury) [inline articles/flow/security/enabling-security.adoc:30]:
```suggestion

Annotate the view classes with the [annotationname]`@AnonymousAllowed`, [annotationname]`@PermitAll`, or [annotationname]`@RolesAllowed` annotations.

```

File: articles/flow/security/enabling-security.adoc [hunk 10]
```diff
@@ -1,358 +1,338 @@
...
+        <artifactId>spring-boot-starter-security</artifactId>
+    </dependency>
+    <!-- other dependencies -->
+</dependencies>
+----
+
+== Log-Out Capability
+
+To handle log-out in web applications, one typically uses a log-out button.
+Here is a basic implementation of log-out button shown on the header of the main layout:
```
Comment (dougjewsbury) [inline articles/flow/security/enabling-security.adoc:97]:
```suggestion
Here is a basic implementation of a log-out button shown on the header of the main layout:

```

File: articles/flow/security/advanced-topics/architecture.adoc (hunk 2 above)
Comment (tarekoraby) [reply articles/flow/security/advanced-topics/architecture.adoc:46]: "Like" is indeed what's intended here.

File: articles/flow/security/enabling-security.adoc (hunk 9 above)
Comment (tarekoraby) [reply articles/flow/security/enabling-security.adoc:31]: @dougjewsbury Why remove the bullet point before this sentence? It is intended to be one of the steps needed to enable the mechanism in a Flow app.

File: articles/flow/security/advanced-topics/security-practices.adoc (hunk 1 above)
Comment (tarekoraby) [reply articles/flow/security/advanced-topics/security-practices.adoc:17]: Thanks for the suggestion @knoobie 👍

File: articles/flow/security/enabling-security.adoc (hunk 9 above)
Comment (dougjewsbury) [reply articles/flow/security/enabling-security.adoc:31]: It was just for grammatical reasons. The list of steps is introduced with "add the following to the project", so "Annotate the classes..." doesn't really fit in. Perhaps you could change "add the following to the project" by "perform these steps", and then insert "Add..." into the steps that require it?

Comment (tarekoraby) [reply articles/flow/security/enabling-security.adoc]: I see. I just made a change that keeps "add the following to the project" as it is, and changes the last step to "One of the following annotations on each view class: [annotationname]`@AnonymousAllowed`, [annotationname]`@PermitAll`, or [annotationname]`@RolesAllowed`."

**Pull request merged**